sxg_rs/src/http_parser/cache_control.rs

```diff
@@ -0,0 +1,123 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use nom::{
+    IResult,
+    branch::alt,
+    bytes::complete::tag_no_case,
+    character::complete::char,
+    combinator::map,
+    combinator::map_res,
+    combinator::opt,
+    sequence::pair,
+    sequence::preceded,
+};
+use std::time::Duration;
+use super::base::{
+    parameter_value,
+    token,
+};
+
+// https://datatracker.ietf.org/doc/html/rfc7234#section-5.2
+#[derive(Clone, Debug, PartialEq)]
+pub enum Directive {
+    SMaxAge(Duration),
+    MaxAge(Duration),
+    // Not relevant to the freshness lifetime computation.
+    Other,
+}
+
+pub fn directive(input: &str) -> IResult<&str, Directive> {
+    alt((
+        // Nonnegative integers up to 31 bits must be parseable per
+        // https://datatracker.ietf.org/doc/html/rfc7234#section-1.2.1.
+        // Parsers may allow higher numbers.
+        preceded(tag_no_case("s-maxage="),
+                 map(map_res(parameter_value, |s| s.parse::<u32>()),
+                     |i| Directive::SMaxAge(Duration::from_secs(i.into())))),
+        preceded(tag_no_case("max-age="),
+                 map(map_res(parameter_value, |s| s.parse::<u32>()),
+                     |i| Directive::MaxAge(Duration::from_secs(i.into())))),
+        map(pair(token, opt(pair(char('='), parameter_value))), |_| Directive::Other),
+    ))(input)
+}
+
+// Returns the freshness lifetime for use in a shared cache, as defined by
+// https://datatracker.ietf.org/doc/html/rfc7234#section-4.2.1, excluding
+// handling of the Expires header. The caller must already have validated the
+// response may be stored in the cache per
+// https://datatracker.ietf.org/doc/html/rfc7234#section-3. 'None' means the
+// lifetime is implicit.
+pub fn freshness_lifetime(directives: Vec<Directive>) -> Option<Duration> {
+    // RFC7234 does not specify what to do if multiple directives of the same
+    // name are present, but it appears that Chromium takes the first:
+    // https://source.chromium.org/chromium/chromium/src/+/main:net/http/http_response_headers.cc;l=799;drc=0ce9df69ba9e32bafc53c3d90db8a707c243da40
+    let mut s_maxage = None::<Duration>;
+    let mut max_age = None::<Duration>;
+    for directive in directives {
+        match directive {
+            Directive::SMaxAge(duration) => { s_maxage.get_or_insert(duration); },
+            Directive::MaxAge(duration) => { max_age.get_or_insert(duration); },
+            Directive::Other => (),
+        };
+    }
+    if let Some(duration) = s_maxage {
```

Comment:
Maybe use `s_maxage.or(max_age)`.

Reply:
Ah, I was looking for that function. Done.